### PR TITLE
feat(portal): support context in TemplatePortal

### LINKED ISSUE #6310: Align TemplatePortal with ComponentPortal and support context in TemplateRef Dialogs
#### What is the expected behavior?
`TemplatePortal` should accept a context which then allows using the `MdDialogConfig.data` as context for a dialog opened using a `TemplateRef`.

#### What is the current behavior?
`TemplatePortal` does no accepts context.
It does accept `locals` but these were added within the first commit and never used.

#### What is the use-case or motivation for changing an existing behavior?
There are several...

First, `TemplatePortal` is not aligned with `ComponentPortal`. Since both provide the same logical functionality it makes sense to return the same logical object: `ComponentRef`, `EmbeddedViewRef`. In addition, `TemplatePortal` does not allow sending data (context) to the template. There is a place holder (`locals`) property but it is never used. `TemplatePortal` is not responsible for the structure of the context so `locals` as `Map<string, any>` is not a proper type, instead a generic type is more suited. The consumer of `TemplatePortal` should decide about the structure (e.g. using `$implicit`? what local to expose etc...)

Once `TemplatePortal` support context, `MdDialog` can provide same experience for both **Component** based dialogs and `TemplateRef` based dialogs.

`MdDialog` will auto crete the context have the `MdDialogConfig.data` as `$implicit` and an extra property `dialogRef` exposed (which holds `MdDialogRef`).

This create a powerful combination of a template that has access to injected data (context) and local template variables of its **parent** template.  
It allows passing template references throughout the application with the ability to have data injected. It's a more declarative approach with a lot of power.

For example, using the `demo-app` code, **without changing TS code** we can provide the same functionality (almost) that the **demo-jazz-dialog** component provide just by changing the html only.

```html
<ng-template let-data let-dialogRef="dialogRef">
  I'm a template dialog. I've been opened {{numTemplateOpens}} times!

  <p>It's Jazz!</p>

  <md-input-container>
    <input mdInput placeholder="How much?" #howMuch>
  </md-input-container>

  <p> {{ data.message }} </p>
  <button type="button" (click)="dialogRef.close(lastCloseResult = howMuch.value)">Close dialog</button>
  <button (click)="dialogRef.updateSize('500px', '500px').updatePosition({ top: '25px', left: '25px' });">Change dimensions</button>`
</ng-template>
```

This template uses bindings from 3 sources, the context, local template variables of the parent component and the component instance itself.

With simple structured directives it will be possible to create template driven UI blockers, modal etc... that accepts data and are set elsewhere.

 I'v started to work on this change, [see the diff](https://github.com/angular/material2/compare/master...shlomiassaf:refactor-TemplatePortal?expand=1)

The change is not big but having TemplatePortal now with a generic required some mods, and it's also a breaking change so it's needs discussion.

Please comment.

cc @jelbourn 
Thanks.

### PR DESCRIPTION
**Pull request by shlomiassaf**
Closes #6310

cc @jelbourn

### CODE REVIEW

File: src/cdk/portal/portal.ts [hunk 1]
```diff
@@ -103,42 +104,38 @@ export class ComponentPortal<T> extends Portal<ComponentRef<T>> {
...
   }
 
   get origin(): ElementRef {
     return this.templateRef.elementRef;
   }
 
-  attach(host: PortalHost, locals?: Map<string, any>): Map<string, any> {
-    this.locals = locals == null ? new Map<string, any>() : locals;
+  attach(host: PortalHost, context?: C): C {
+    this.context = context!;
```
Comment (jelbourn) [inline src/cdk/portal/portal.ts]: Why use the `!` here? The class property is marked as `| undefined` so it shouldn't be necessary

File: src/lib/dialog/dialog.ts [hunk 2]
```diff
@@ -228,7 +228,9 @@ export class MdDialog {
     }
 
     if (componentOrTemplateRef instanceof TemplateRef) {
-      dialogContainer.attachTemplatePortal(new TemplatePortal(componentOrTemplateRef, null!));
+      dialogContainer.attachTemplatePortal(
+        new TemplatePortal<T>(componentOrTemplateRef, null!,
```
Comment (jelbourn) [inline src/lib/dialog/dialog.ts]: Doesn't `null!` resolve to `never`? Seems like this needs a real `ViewContainerRef`

File: src/lib/snack-bar/snack-bar-container.ts [hunk 3]
```diff
@@ -110,7 +111,8 @@ export class MdSnackBarContainer extends BasePortalHost implements OnDestroy {
   }
 
   /** Attach a template portal as content to this snack bar container. */
-  attachTemplatePortal(): Map<string, any> {
+  /* tslint:disable-next-line */
```
Comment (jelbourn) [inline src/lib/snack-bar/snack-bar-container.ts]: What's the lint problem here?

File: src/cdk/portal/portal.spec.ts [hunk 4]
```diff
@@ -299,15 +299,15 @@ describe('Portals', () => {
       fixture.detectChanges();
 
       // Attach the TemplatePortal.
-      testAppComponent.portalWithBinding.attach(host);
+      testAppComponent.portalWithBinding.attach(host, { $implicit: { status: 'fresh' } });
```
Comment (jelbourn) [inline src/cdk/portal/portal.spec.ts]: Omit spaces in braces
```ts
{$implicit: {status: 'fresh'}}
```

File: src/cdk/portal/portal.ts [hunk 5]
```diff
@@ -103,42 +104,38 @@ export class ComponentPortal<T> extends Portal<ComponentRef<T>> {
...
-   * Additional locals for the instantiated embedded view.
-   * These locals can be seen as "exports" for the template, such as how ngFor has
-   * index / event / odd.
-   * See https://angular.io/docs/ts/latest/api/core/EmbeddedViewRef-class.html
-   */
-  locals: Map<string, any> = new Map<string, any>();
+  context: C | undefined;
 
-  constructor(template: TemplateRef<any>, viewContainerRef: ViewContainerRef) {
+  constructor(template: TemplateRef<any>, viewContainerRef: ViewContainerRef, context?: C) {
```
Comment (jelbourn) [inline src/cdk/portal/portal.ts]: It's somewhat strange that the `context` can be set either in the constructor or in `attach`. Is this meant to be a _default_ context when nothing is passed to `attach`?

File: src/cdk/portal/portal.ts [hunk 6]
```diff
@@ -103,42 +104,38 @@ export class ComponentPortal<T> extends Portal<ComponentRef<T>> {
...
     super();
     this.templateRef = template;
     this.viewContainerRef = viewContainerRef;
+    if (context) {
+     this.context = context;
+    }
```
Comment (jelbourn) [inline src/cdk/portal/portal.ts]: You should just be able to put `public context?: C` in the constructor and omit this assignment

File: src/cdk/portal/portal.ts (hunk 1 above)
Comment (jelbourn) [inline src/cdk/portal/portal.ts]: Won't this always override the context given in the constructor, even if no context is passed here? You should be able to do something like
`attach(host, PortalHost, context: C = this.context)`

File: src/cdk/portal/portal.spec.ts (hunk 4 above)
Comment (jelbourn) [inline src/cdk/portal/portal.spec.ts]: There should be:
* One test where no context is given
* One test where context is given in the constructor
* One test where context is given in `attach`
* One test where context is given in both the constructor and in attach

File: src/lib/dialog/dialog.spec.ts [hunk 7]
```diff
@@ -80,6 +81,28 @@ describe('MdDialog', () => {
     expect(dialogContainerElement.getAttribute('role')).toBe('dialog');
   });
 
+  it('should open a dialog with a template', () => {
+    const templateRefFixture = TestBed.createComponent(ComponentWithTemplateRef);
+    templateRefFixture.componentInstance.localValue = 'Bees';
+    templateRefFixture.detectChanges();
+
+    const data = { value: 'Knees' };
```
Comment (jelbourn) [inline src/lib/dialog/dialog.spec.ts]: Omit spaces inside braces

File: src/cdk/portal/portal.ts (hunk 5 above)
Comment (shlomiassaf) [inline src/cdk/portal/portal.ts]: This is required to support the abstraction in [dialog.ts](https://github.com/shlomiassaf/material2/blob/9ee6478063fd081b824d206abc93ebe803947d43/src/lib/dialog/dialog.ts#L232-L233)

If you follow the trail, the `attach` method is never called and `BasePortalHost` (`PortalHostDirective) handles it from the top.

Adding a context to the method call of `attachTemplatePortal` is a mess since it involves changing a lot of classes that extend `BasePortal`, for example `MdDialogContainer`

Additionally, changing the signature of `attachTemplatePortal` to accept a context will make it different from the signature `attachComponentPortal`...

File: src/cdk/portal/portal.ts (hunk 6 above)
Comment (shlomiassaf) [inline src/cdk/portal/portal.ts]: I just tried keeping the code style the same.

The other 2 parameters (`template` and `template`) are also public but they are manually assigned without using TypeScript's constructor parameter modifiers....

So I just followed that.

File: src/cdk/portal/portal.ts (hunk 1 above)
Comment (shlomiassaf) [inline src/cdk/portal/portal.ts]: Indeed.

I think `this.context = context || this.context;` is more readable but I used your approach

It requires a slight modification to be friends with typescript:
```
attach(host: PortalHost, context: C | undefined = this.context): C {
```

File: src/lib/dialog/dialog.ts (hunk 2 above)
Comment (shlomiassaf) [inline src/lib/dialog/dialog.ts]: This was the logic before I made the change so I left it as is.

I think if a change is required here it should be handled in a different PR not to mixup things.

It does work...

File: src/lib/snack-bar/snack-bar-container.ts (hunk 3 above)
Comment (shlomiassaf) [inline src/lib/snack-bar/snack-bar-container.ts]: 'C' is declared but never used.

File: src/lib/dialog/dialog.spec.ts (hunk 7 above)
Comment (shlomiassaf) [inline src/lib/dialog/dialog.spec.ts]: 👍

File: src/cdk/portal/portal.spec.ts (hunk 4 above)
Comment (shlomiassaf) [inline src/cdk/portal/portal.spec.ts]: 👍

File: src/cdk/portal/portal.ts (hunk 1 above)
Comment (shlomiassaf) [inline src/cdk/portal/portal.ts]: 👍

File: src/cdk/portal/portal.spec.ts (hunk 4 above)
Comment (shlomiassaf) [inline src/cdk/portal/portal.spec.ts]: Actually the tests are limited since there are no are no "low-level" tests for `TemplatePortal` with a `PortalHostDirective` (there are some for `ComponentPortal`)

I'v added 2 tests:
  1. **should load a template into the portal**
Check basic projection of programatically created `TemplatePortal`

  2. **should project template context bindings in the portal**
Check project of programatically created `TemplatePortal` with 4 internal `expects`
       - No context
       - Context through `attach()` method
       - Context through **constructor**
       - Context through **constructor** and through `attach()` method - the latter must take precedence.

File: src/cdk/portal/portal.ts (hunk 6 above)
Comment (jelbourn) [inline src/cdk/portal/portal.ts]: Ah, I think that's just from this being one of the first things I wrote in the library and not being as familiar with TypeScript back then.

File: src/cdk/portal/portal.ts (hunk 5 above)
Comment (jelbourn) [inline src/cdk/portal/portal.ts]: I see what you mean. I think it's okay to include it in the signature, then, so long as the JsDoc makes it clear that this updates the `context` property of the `TemplatePortal`

File: src/lib/snack-bar/snack-bar-container.ts (hunk 3 above)
Comment (jelbourn) [inline src/lib/snack-bar/snack-bar-container.ts]: Ah, you should be able to do 
```ts
attachTemplatePortal(): EmbeddedViewRef<any> {
  ...
}
```

Comment (shlomiassaf) [inline src/lib/snack-bar/snack-bar-container.ts]: Yes, we can do that but this will not be inline with other `attachTemplatePortal` calls that does accept a type parameter.

This is not the only reason, main reason is that right now we need to add the `tslint disable` but once we move to TypeScript 2.4 it will be easy to migrate it to:

```ts
attachTemplatePortal<C = any>(): EmbeddedViewRef<C> {
      throw Error('Not yet implemented');
}
```

This way we allow setting the type parameter for those who want.
Later we will support those who "doesnt" want as well :)

Comment (jelbourn) [inline src/lib/snack-bar/snack-bar-container.ts]: Using `any` here right now is the most technically correct thing, though (until the function is actually implemented). The linter is catching a real issue in that the function genuinely isn't generic.

Comment (shlomiassaf) [inline src/lib/snack-bar/snack-bar-container.ts]: No problem, will update it

File: src/cdk/portal/portal.ts (hunk 5 above)
Comment (shlomiassaf) [inline src/cdk/portal/portal.ts]: @jelbourn I'v added a JsDoc comment to the `attach` method acoordignly.

I didn't find an example of how to comment a constructor signature, and if its like commenting a method, so I left it as is.

FYI: Both `template` and  `viewContainerRef` constructor parameters do the same thing (set a property on the instance) and are not commented so I don't think we should go into it. It is self explaining.

**Comment by shlomiassaf** [comment]
@jelbourn Why does it fail?

**Comment by jelbourn** [comment]
e2e tests are just timing out, which happens sometimes
